src/image_bios.cpp

```diff
@@ -198,15 +202,59 @@ void BIOSUpdater::unlock()
 
 void BIOSUpdater::doInstall(const fs::path& file)
 {
-    constexpr auto mtd = "/dev/mtd/bios";
-
     // NOTE: This process may take a lot of time and we want to show the
     //       progress from original flashcp output.
-    printf("Writing %s to %s\n", file.filename().c_str(), mtd);
-    int rc = system(strfmt("flashcp -v %s %s", file.c_str(), mtd).c_str());
+    printf("Writing %s to %s\n", file.filename().c_str(), mtdDevice);
+    int rc =
+        system(strfmt("flashcp -v %s %s", file.c_str(), mtdDevice).c_str());
     checkWaitStatus(rc, "");
 }
 
+void BIOSUpdater::doBeforeInstall(bool reset)
+{
+    if (!reset)
+    {
+        puts("Preserving BIOS settings...");
+        const fs::path dumpFile(tmpdir / nvramFile);
+        const std::string cmd =
+            strfmt("nanddump --startaddress %lu --length %lu --file '%s' %s",
+                   nvramOffset, nvramSize, dumpFile.c_str(), mtdDevice);
+        const int rc = system(cmd.c_str());
+        checkWaitStatus(rc, std::string());
+        if (!fs::exists(dumpFile))
+        {
+            throw FwupdateError("Error reading NVRAM");
```

Comment:
I have a doubt. Is it really the fatal error.
Probably we could ask the users opinion and continue without preserving. The same way was implemented for the OpenPOWER firmware. 
What do you think?

Reply:
I think it is a critical error, end-user expects preserving BIOS settings but we can't do it.
I am not sure we can use confirmation here. It is an error, so it cannot be silently ignored with `-y` option.

Reply:
Ok, let it be as is.